packages/apollo-engine-reporting/src/agent.ts

```diff
@@ -325,12 +395,41 @@ export class EngineReportingAgent<TContext = any> {
       });
     }
 
+    if (this.options.endpointUrl) {
+      this.logger.warn(
+        '[deprecated] The `endpointUrl` option within `engine` has been renamed to `tracesEndpointUrl`.',
+      );
+    }
+    this.tracesEndpointUrl =
+      (this.options.endpointUrl ||
+        this.options.tracesEndpointUrl ||
+        'https://engine-report.apollodata.com') + '/api/ingress/traces';
+
     // Handle the legacy options: privateVariables and privateHeaders
     handleLegacyOptions(this.options);
   }
 
+  private executableSchemaIdGenerator(schema: string | GraphQLSchema) {
+    if (this.lastSeenExecutableSchemaToId?.executableSchema === schema) {
+      return this.lastSeenExecutableSchemaToId.executableSchemaId;
+    }
+    const id = computeExecutableSchemaId(schema);
+
+    // We override this variable every time we get a new schema so we cache
+    // the last seen value. It mostly a cached pair.
+    this.lastSeenExecutableSchemaToId = {
+      executableSchema: schema,
+      executableSchemaId: id,
+    };
+
+    return id;
+  }
+
   public newPlugin(): ApolloServerPlugin<TContext> {
-    return plugin(this.options, this.addTrace.bind(this));
+    return plugin(this.options, this.addTrace.bind(this), {
+      startSchemaReporting: this.startSchemaReporting.bind(this),
```

Comment:
Now that there are so many callbacks, I'd say just define an interface `EngineReportingAgentPluginAPI` that has those three methods on it and just passing `this` to the plugin. Or just pass `this: EngineReportingAgent` and don't stress too much about defining the API to be smaller.